presto-raptorx/src/main/java/com/facebook/presto/raptorx/storage/OrcFileRewriter.java

```diff
@@ -0,0 +1,194 @@
+import java.util.BitSet;
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.raptorx.RaptorErrorCode.RAPTOR_STORAGE_ERROR;
+import static com.facebook.presto.raptorx.util.OrcUtil.createOrcFileWriter;
+import static com.facebook.presto.raptorx.util.OrcUtil.createOrcReader;
+import static com.facebook.presto.raptorx.util.OrcUtil.createOrcRecordReader;
+import static com.facebook.presto.raptorx.util.OrcUtil.fileOrcDataSource;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.units.Duration.nanosSince;
+import static java.lang.Long.parseLong;
+import static java.lang.Math.toIntExact;
+
+public final class OrcFileRewriter
+{
+    private static final Logger log = Logger.get(OrcFileRewriter.class);
+
+    private OrcFileRewriter() {}
+
+    public static OrcFileInfo rewriteOrcFile(
+            long newChunkId,
+            File input,
+            File output,
+            Map<Long, Type> columns,
+            CompressionType compressionType,
+            BitSet rowsToDelete,
+            ReaderAttributes readerAttributes,
+            TypeManager typeManager,
+            OrcWriterStats stats)
+    {
+        long start = System.nanoTime();
+        try (OrcDataSource dataSource = fileOrcDataSource(input, readerAttributes)) {
+            OrcReader reader = createOrcReader(dataSource, readerAttributes);
+
+            ImmutableMap.Builder<Integer, Type> columnBuilder = ImmutableMap.builder();
+            ImmutableList.Builder<Long> columnIdBuilder = ImmutableList.builder();
+            for (int i = 0; i < reader.getColumnNames().size(); i++) {
+                long columnId = parseLong(reader.getColumnNames().get(i));
+                Type type = columns.get(columnId);
+                if (type != null) {
+                    columnBuilder.put(i, type);
+                    columnIdBuilder.add(columnId);
+                }
+            }
+            Map<Integer, Type> includedColumns = columnBuilder.build();
+            List<Long> columnIds = columnIdBuilder.build();
+            List<Type> columnTypes = includedColumns.values().stream().collect(toImmutableList());
+
+            StorageTypeConverter converter = new StorageTypeConverter(typeManager);
+            List<Type> storageTypes = columnTypes.stream()
+                    .map(converter::toStorageType)
+                    .collect(toImmutableList());
+
+            OrcRecordReader recordReader = createOrcRecordReader(reader, includedColumns);
+
+            long fileRowCount = recordReader.getFileRowCount();
+            if (fileRowCount < rowsToDelete.length()) {
+                throw new IOException("File has fewer rows than deletion vector");
+            }
+            int deleteRowCount = rowsToDelete.cardinality();
+            if (fileRowCount == deleteRowCount) {
+                return new OrcFileInfo(0, 0);
+            }
+            if (fileRowCount >= Integer.MAX_VALUE) {
```

Comment:
"highker on Feb 4  Member
Why we limit the row count?"

Since during write, we never exceed "Integer.MAX_VALUE", so here during rewrite is may not be needed. But I will still keep it here, same as V1, just for unusual cases like writing has bug, or disk file changed.

And the reason we have limit during write, and together with SIZE_LIMIT, is just for limit each file's rows and size.